enhancements/cluster-api/crd-compatibility-checker.md

```diff
@@ -0,0 +1,469 @@
+  However, it would require further architectural work to deploy and maintain with OCP.
+  We should revisit this should it ever become feasible in the future.
+
+- **Manual coordination**:
+  Rely on out-of-band coordination between actors.
+  
+  This was considered unlikely to be a robust process over time.
+
+## Open Questions [optional]
+
+1. What is the specific algorithm for determining CRD compatibility?
+2. How should the system handle CRD conversion webhooks during compatibility checks?
+3. What is the performance impact of the admission webhooks on high-traffic clusters?
+
+## Test Plan
+
+**Note:** *Section not required until targeted at a release.*
+
+The test plan should include:
+- Unit tests for compatibility checking logic
+- Integration tests for webhook admission scenarios
+- E2E tests for complete workflows including upgrade scenarios
+- Performance tests to measure webhook latency impact
+- Tests for various failure modes and recovery scenarios
+
+## Graduation Criteria
+
+**Note:** *Section not required until targeted at a release.*
+
+### Dev Preview -> Tech Preview
+
+- Webhook admission working for CRD updates
+- Feature integrated in cluster-capi-operator
+- Initial documentation and API stability
+- Comprehensive test coverage for all deployed features
+
+### Tech Preview -> GA
+
+- Object schema validation working end-to-end
+- Comprehensive test coverage including upgrade/downgrade scenarios
+- Performance validation showing acceptable webhook latency
+- Production deployment validation
+- Complete user documentation
+
+## Upgrade / Downgrade Strategy
+
+CRDCompatibilityChecker has no direct impact on cluster operations unless it is configured.
+We handle upgrade and downgrade by altering the behaviour of cluster-capi-operator, which is responsible for the in-payload configuration.
+
+The following requires 2 feature gates:
+* CRDCompatibilityCheckerUpgradeCheck
+* CRDCompatibilityCheckerEnforce
+
+`CRDCompatibilityCheckerUpgradeCheck` enables the 'future version' check.
+If `CRDCompatibilityCheckerUpgradeCheck` is enabled, cluster-capi-operator will:
+
+* Create a `CRDCompatibilityRequirement` for every CRD discovered in the 'future' transport config map if that CRD is also marked 'unmanaged'.
+This requirement will have `crdAdmitAction=Warn`.
+* Mark itself not upgradeable if any of these are marked `Compatible=False`.
+
+If `CRDCompatibilityCheckerUpgradeCheck` is not enabled, cluster-capi-operator will look for all 'future' `CRDCompatibilityRequirement`s and delete them.
+
+If `CRDCompatibilityCheckerEnforce` is enabled, cluster-capi-operator will:
+* Create a `CRDCompatibilityRequirement` for every CRD in the 'active' transport config map if that CRD is also marked 'unmanaged'.
+This requirement will have `crdAdmitAction=Enforce`.
+
+If `CRDCompatibilityCheckerEnforce` is not enabled, cluster-capi-operator will look for all 'active' `CRDCompatibilityRequirement`s and delete them.
```

Comment:
We don't generally both with disabled logic like this, once a gate is enabled in OCP, it should never be disabled

Reply:
I've specifically documented this behaviour because it neatly covers the downgrade case, and I'm guessing it's probably not too hard to implement. Happy to take it out, but I think we'd spend as much effort writing the documentation on how to do it manually.

Reply:
I've noted that we don't intend to support downgrades.